Entity cluster refactor

erikmaarten opened the pull request:
Fixes #4499. EntityCluster now avoids globbing properties on entities, and keeps those properties (collection indices) in an object.

There were three functions here that did the same thing but for different collections, so I made a factory function that combines the functionality. It could use a bit of polish and perhaps a better name, but please have a look at this first and see what you think.

erikmaarten commented:
@mramato I refactored this a bit while I was at it -- what do you think?

mramato commented on Source/DataSources/EntityCluster.js:
```diff
@@ -87,6 +87,8 @@ define([
         this._clusterLabelCollection = undefined;
         this._clusterPointCollection = undefined;
 
+        this._collectionIndicesByEntity = Object.create(null);
```
`Object.create` is actually a performance bottleneck so as a general rule we don't use it in our code.  Just use `{}` instead.

mramato commented on Source/DataSources/EntityCluster.js:
```diff
@@ -527,43 +529,55 @@ define([
+            var entityIndices = context._collectionIndicesByEntity[entity.id];
+
+            if (!defined(entityIndices)) {
+                entityIndices = context._collectionIndicesByEntity[entity.id] = Object.create(null);
```
Since we know for a fact that this object will have three properties, pointIndex, labelIndex, and billboardIndex, we should specify an object with those three properties here.  It will let the browser make additional optimizations that it otherwise couldn't do.

mramato commented:
One potential issue I see here is that `_collectionIndicesByEntity` never shrinks (unless I'm missing something).  We probably want to call `delete this._collectionIndicesByEntity[id]` in the 3 `.removeXXX` functions if none of the 3 properties on the object are defined.  If the overhead on that proves problematic, we can probably do something fancier, but let's start there.

I'm going to run a quick performance comparison here as well, just to make sure we're not introducing overhead with the property index lookups, but I don't expect it to be a problem.

Thanks @erikmaarten!

mramato commented:
@erikmaarten can you take a look at #4409 next?  I added it to your list of issues; it should be an easy fix since it's related.

mramato commented:
Running http://localhost:8080/Apps/Sandcastle/index.html?src=Clustering.html&label=Showcases in this branch seems to cause a crash (works in master).  Can we fix it and add a unit test to make sure that case is covered.  Thanks.

twpayne commented:
FWIW the following commits contain unit tests that might be useful:

https://github.com/AnalyticalGraphicsInc/cesium/pull/4482/commits/299bcc5d86767a18601d63e1bbafd682a5579ecf
https://github.com/AnalyticalGraphicsInc/cesium/pull/4482/commits/99c04066de28443d0c2b53ba557a54fb9a442220
https://github.com/AnalyticalGraphicsInc/cesium/pull/4482/commits/74b32aebbd290edb07d0b633337b81dd474f88aa

All of these unit tests fail on `master` at the time of writing and should pass once #4409 is fixed.

erikmaarten commented:
@mramato Thanks for the review. I've addressed the issues, and like you suggested I'm now deleting entries from `_collectionIndicesByEntity`. I ran into an issue where two visualizer tests failed and wanted to make sure my removal function did what I expected I added some simple tests. The issue turned out to be that removeXXX is sometimes run before XXX is actually created, so I added a check to see whether `_.collectionIndicesByEntity[id]` exists before removing.

erikmaarten commented:
@mramato Looking at the Sandcastle crash, will add another commit with a fix.

erikmaarten commented:
@mramato Fixed the problem appearing in the Sandcastle example. Covering this with a test turned out a little tricky, the code in question has a few different branches and conditions that affect when `getBoundingBox` and `addNonClusteredItem` get called. In the end I realized that when an item has an associated label (`item.id.label`) the problem is triggered, so I made test case identical to the `clusters points` except that a point has a label.

erikmaarten commented:
There's another issue here, `item.id` doesn't seem to be the actual id, but an object. Will fix this before this is ready.

erikmaarten commented:
Thanks @twpayne.

@mramato This is ready. I've also addressed #4409 in this PR, since it's affected by previous changes here.

mramato commented:
Awesome work here @erikmaarten!  Thanks.
